server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexAction.java

```diff
@@ -0,0 +1,653 @@
+        }
+
+        public List<ResolvedIndex> getIndices() {
+            return indices;
+        }
+
+        public List<ResolvedAlias> getAliases() {
+            return aliases;
+        }
+
+        public List<ResolvedDataStream> getDataStreams() {
+            return dataStreams;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeVInt(indices.size());
+            for (ResolvedIndex index : indices) {
+                index.writeTo(out);
+            }
+            out.writeVInt(aliases.size());
+            for (ResolvedAlias alias : aliases) {
+                alias.writeTo(out);
+            }
+            out.writeVInt(dataStreams.size());
+            for (ResolvedDataStream dataStream : dataStreams) {
+                dataStream.writeTo(out);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(INDICES_FIELD.getPreferredName(), indices);
+            builder.field(ALIASES_FIELD.getPreferredName(), aliases);
+            builder.field(DATA_STREAMS_FIELD.getPreferredName(), dataStreams);
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Response response = (Response) o;
+            return indices.equals(response.indices) && aliases.equals(response.aliases) && dataStreams.equals(response.dataStreams);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(indices, aliases, dataStreams);
+        }
+    }
+
+    public static class TransportAction extends HandledTransportAction<Request, Response> {
+
+        private final ThreadPool threadPool;
+        private final ClusterService clusterService;
+        private final RemoteClusterService remoteClusterService;
+        private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+        @Inject
+        public TransportAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+            super(NAME, transportService, actionFilters, ResolveIndexAction.Request::new);
+            this.threadPool = threadPool;
+            this.clusterService = clusterService;
+            this.remoteClusterService = transportService.getRemoteClusterService();
+            this.indexNameExpressionResolver = indexNameExpressionResolver;
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, final ActionListener<Response> listener) {
+            final ClusterState clusterState = clusterService.state();
```

Comment:
Maybe extract the logic here in a static method, so that unit tests can be written?